### PR TITLE
Outdoors map style support

### PR DESCRIPTION
**Pull request by pastk**
Adds a compiled outdoors map style (changes and screenshots are same as in #2804) into android apk.
There is no UI switcher yet.
To switch into outdoors use a search command `?olight`, issue `?light` to switch back to normal style.
The style will be applied full only after maps regeneration (due to features' zoom level visibility changes). Or compile with a patch #2798 to be able to test the full style without maps regen.

Tested on Android and desktop only.
Might break iOS (as new style files should be added to the project files, etc.) - need help here!

TODO (probably best in separate PRs):
- add Android UI style switcher - anyone wants to do it? @arnaudvergnet maybe? As you worked with the layers button recently
- add iOS UI style switcher
- add desktop UI style switcher
- regenerate maps and check impact on size
- polish the new style to make it release worthy
- refactor style files handling in core/scripts to move from hardcoded to more dynamic way based on existing style files/folders - to make it easier to add new core styles, add custom styles, support different style files for different data versions

Needed-for: https://github.com/organicmaps/organicmaps/issues/2356

### CODE REVIEW

File: data/styles/outdoors/include/Basemap_label.mapcss [hunk 1]
```diff
@@ -0,0 +1,764 @@
+/* ~~~~ CONTENT OF BASEMAP lable ~~~~~
+
+1.Z-INDEX of BASEMAP LABEL
```
Comment (biodranik) [inline data/styles/outdoors/include/Basemap_label.mapcss:3]: Эти файлы полностью дублируют остальные стили? Хотелось бы иметь максимально общую базу, и в разных стилях хранить только оверрайды базы, иначе замучаемся с поддержкой, добавив ещё 3-4 стиля.

File: map/framework.cpp [hunk 2]
```diff
@@ -2576,10 +2576,14 @@ bool Framework::ParseDrapeDebugCommand(string const & query)
     desiredStyle = MapStyleDark;
   else if (query == "?light" || query == "mapstyle:light")
     desiredStyle = MapStyleClear;
-  else if (query == "?vdark" || query == "mapstyle:vehicle_dark")
-    desiredStyle = MapStyleVehicleDark;
   else if (query == "?vlight" || query == "mapstyle:vehicle_light")
     desiredStyle = MapStyleVehicleClear;
+  else if (query == "?vdark" || query == "mapstyle:vehicle_dark")
+    desiredStyle = MapStyleVehicleDark;
+  else if (query == "?outd-light" || query == "mapstyle:outdoors_light")
```
Comment (biodranik) [inline map/framework.cpp:2583]: olight/odark?

**Review by biodranik — COMMENTED** [review]
На сколько увеличился размер данных в этом PR по сравнению с мастером, на примере какой-то отдельной страны/региона, где много хайков?

File: data/styles/outdoors/include/Basemap_label.mapcss (hunk 1 above)
Comment (pastk) [reply data/styles/outdoors/include/Basemap_label.mapcss:3]: Yeap, its in plans.

**Comment by pastk** [comment]
> На сколько увеличился размер данных в этом PR по сравнению с мастером, на примере какой-то отдельной страны/региона, где много хайков?

I don't know, as I don't have a map gen pipeline setup.
It would be better (much quicker at least) if someone else could regen maps with a new style and check.

File: data/styles/outdoors/include/Basemap_label.mapcss (hunk 1 above)
Comment (biodranik) [reply data/styles/outdoors/include/Basemap_label.mapcss:3]: Maybe we can start with these plans first, to avoid unnecessary copy-paste?

Comment (pastk) [reply data/styles/outdoors/include/Basemap_label.mapcss:3]: Yeap I'm going to try it asap.

**Comment by vng** [comment]
Suspicious that first commit https://github.com/organicmaps/organicmaps/pull/4247/commits/bff0ea9004ecd169edf175268566bd06dc1e71dd, adding outdoor style switch into the code has data/colors.txt modifications like removing a bunch of colors ..

I suspect that this file should be in the regenerated commit, or something goes wrong ..

**Comment by pastk** [comment]
Made *.mapcss files of the new style contain only changes to the base style.
I had to copy some "eval" sections still due to #4258.

Also copied some unchanged rules in case they belong to a logical block when the most of it was changed, e.g. isolines (all rules are changed somewhat except `[isoline=step_10]`).

**Comment by biodranik** [comment]
Is there a way to "validate" mapcss and report some issues/fail on errors if necessary?

**Comment by pastk** [comment]
> Is there a way to "validate" mapcss and report some issues/fail on errors if necessary?

I think we don't have such a tool. @vng ?
TBH our style files are full of duplicate and overriding rules, not working rules etc.

**Comment by pastk** [comment]
Tested it on desktop - works well.

**Comment by pastk** [comment]
Removed copy-pasted eval()s sections.

**Comment by pastk** [comment]
I suggest to merge it and then we can continue working on it.
After the merge and maps regeneration the outdoors style will be available for beta-testing (switching via search commands - we can tell about them to selected people or just publish in the chat).

**Comment by vng** [comment]
Run generator on your branch. Lets see what will we get.

**Comment by vng** [comment]
Test here https://github.com/organicmaps/organicmaps/pull/4399

@pastk Should we also move isolines on earlier scales?

**Comment by biodranik** [comment]
At which scales are isolines drawn now? Can we draw them at the level where no additional increase in the geometry size will be observed?

**Comment by pastk** [comment]
> At which scales are isolines drawn now? Can we draw them at the level where no additional increase in the geometry size will be observed?

I've increased their visibility by one zoom level only. However in all cases it was a "border" level, so it triggered addition of another generated geom level.
So its either accepting the size increase or leaving isolines visibility unchanged. Maybe its possible to make all isolines more visible except line_50 as it generates the most of extra data, but we'll need to see if it affects relief readability a lot or not.
On the upside - we might use that extra lines density to improve the generic style as well, its not necessary to keep it to outdoors style only.

There is also an option to save some space used by isolines by reviewing params of their generation. IMO they're too detailed/precise (e.g. not much point to have a special line for a small 20-meters long/wide hill).
In my experience density of lines is more important than their precision for an efficient relief reading.

**Comment by vng** [comment]
There are a bunch of isolines generation profiles at ./data/conf/isolines/
Playing with this params needs a huge amount of time and patience.
Probably, if some folks can deploy SRTM data (partially) for his native country, run generator with different isoline profiles and select optimal - sounds like a plan.

**Comment by eddy-geek** [comment]
>  IMO they're too detailed/precise

I haven't seen the the change discussed here, but precise contour lines are very useful for some activities - at least skiing and ski-touring (but of course OM has to find the right balance for general use).

On my side I'd be interested to contribute contour lines based on more precise local elevation data that I'm aggregating (X-ref eslopemap/eslope#3), if it's not too hard to set that up on a laptop.

**Comment by pastk** [comment]
> Test here #4399

So looks like size-wise its allright and switching etc. works as expected.
So I think its good to be merged.
Rebased.

**Comment by biodranik** [comment]
I've added drules files to iOS and rebased it on another branch: https://github.com/organicmaps/organicmaps/pull/new/alex-outdoors

**Comment by biodranik** [comment]
I've checked the style, and it looks ugly now. After a minute of playing with the map I found that:
- Railroads are too aggressive, in cities, they look ugly
- Hiking paths are a bit too bold IMO, they (or at least major ones) should appear earlier
- Many objects pollute the map and make it not too usable (caves, waterfalls)

The style definitely needs some love. Let's check how our focused competitors are doing, like Komoot, AllTrails and others.

**Comment by pastk** [comment]
> Could you please provide screenshots?

https://github.com/organicmaps/organicmaps/pull/2804

**Comment by pastk** [comment]
> I've checked the style, and it looks ugly now. After a minute of playing with the map I found that:
> 
>     * Railroads are too aggressive, in cities, they look ugly

Cities are overall very ugly with an outdoors style :) And that was expected, isn't it?
And it won't be solved really unless we have dynamic styling.
People were asking specifically for much more visible railroads for hiking...

> 
>     * Hiking paths are a bit too bold IMO, they (or at least major ones) should appear earlier
> 

For my taste they're too bold indeed, but people were asking for bold trails specifically so that they're visible in a bright sun...
Appear earlier than z11? Are you sure? ATM there is no way to distinguish major trails, unfortunately - its a separate big work.

>     * Many objects pollute the map and make it not too usable (caves, waterfalls)

Their visibility was risen by 1 zoom level only. And its important waterfalls and caves are more visible in outdoors style than in a regular one. Of course, if you're looking at the Switzerland map, then waterfalls and caves pollute it even in the main style, this is a known issue (and dynamic styling is an answer).

**Comment by pastk** [comment]
> > At which scales are isolines drawn now? Can we draw them at the level where no additional increase in the geometry size will be observed?
> 
> I've increased their visibility by one zoom level only. However in all cases it was a "border" level, so it triggered addition of another generated geom level. So its either accepting the size increase or leaving isolines visibility unchanged. Maybe its possible to make all isolines more visible except line_50 as it generates the most of extra data, but we'll need to see if it affects relief readability a lot or not. On the upside - we might use that extra lines density to improve the generic style as well, its not necessary to keep it to outdoors style only.

1/3 of mwms use `step_50` isolines (which add the most of data). It's possible to do without rising `step_50` isolines, in that case max mwm size increase will be limited to ~5%. But they add to the readability of the map for sure (though not as important as `step_100` visibility raise). So I suggest leaving `step_50` visibility increase as is.

Moreover, the extra data will be useful for the main style too: https://github.com/organicmaps/organicmaps/pull/4580

**Comment by biodranik** [comment]
Do you think that it is a good idea to release an "ugly" hiking style to the public? It doesn't look attractive on screenshots, how will we promote it? :)

So we can either release an "ugly", not-polished style now to satisfy a few existing users or polish the style/fix other important blockers/issues to implement it in a more usable way, by also checking out the styles of our competitors.

Regarding isolines: if they were in a separate downloadable file, then it won't be an issue at all. But now we force everyone, even those who do not need/do not use isolines, to download them and suffer from the increased size of the data. Separating isolines from other maps data will greatly improve UX for many users who don't need them.

@vng what do you think about the refactoring to separate isolines data? After implementing it, we can enable them in those areas where they're disabled now. And then we can independently replace the existing format with another one.

**Comment by pastk** [comment]
> Do you think that it is a good idea to release an "ugly" hiking style to the public? It doesn't look attractive on screenshots, how will we promote it? :)
> 
> So we can either release an "ugly", not-polished style now to satisfy a few existing users or polish the style/fix other important blockers/issues to implement it in a more usable way, by also checking out the styles of our competitors.

My suggestion was only about merging this PR as is, i.e. without the styles switch UI. This way we can attract enthusiast people form the chat to actually use the style and give more feedback and !hopefully! inspire someone to work on improving/polishing it, as I personally just don't have enough time for that atm.

**Comment by biodranik** [comment]
We can track and release beta builds with this style if someone really needs it on a device. Now it's hard to imagine how a contributor will modify the style without building the app locally. Meaning, this branch can live unmerged and updated, like we're doing now with AA.

**Comment by biodranik** [comment]
There are betas that can be easily installed for anyone who wants to get feedback. Adding an easy way to switch between styles would be even better, so users will see it in the beta and can easily compare the same areas.

**Comment by pastk** [comment]
Summary of users' feedback to April outdoors beta:

Requested enhancements:
- [ ] trails stand out too much (due to their boldness) compared to roads, waterways, info & shelter icons, etc.
- [ ] trails are too bold and/or too dense in some cases; #2185 and #5415 will help partially
- [ ] railways are too thick/overlapping at the train stations, #2844
- [ ] forest, shrubs, meadow, field, etc. should look more apart
- [ ] differentiate track types
- [ ] inconsistencies in areal features disappering when zooming out (e.g. man made big features disappear before natural features) - it applies to the main style also; related to #2856
- [ ] mountain peak icons should be bigger and more visible compared to passes/saddles and caves icons
- [ ] add natural=ridge/dune/sinkhole/valley, place=isolated_dwelling/farm, man_made=cairn/tower and other outdoor-related features (e.g. a good list https://github.com/organicmaps/organicmaps/issues/2356#issuecomment-1501720742)

Bugs:
- [ ] sometimes a change of zoom is needed for a style switch to take effect

**Comment by pastk** [comment]
The biggest challenge though is that there are two distinct cases:
- "field" use in harsh conditions, bright sun etc. - requires very high contrast, thick lines, big icons and captions etc.
- trip planning - requires detailed overviews

Ideally we would have two distinct styles or a sub-style, but its challenging also how to present it UX-wise and atm we don't have enough resources to develop and then maintain two new styles.

So realistically in the meantime we can either focus on just one case (e.g. "planning" mode with thinner lines)
or try to find some balance (reduce lines' thickness at higher zooms, but increase color contrast and/or add line casings/halos, etc.)

WDYT?

**Comment by biodranik** [comment]
It is not the first time we see the need to have different UX and styles for planning compared to the "field" use of the app. As our long-term goal is to make the best maps app, then the right direction is to support both "modes", or "styles". You definitely don't need high contrast and bold lines when you plan the route. Let's start collecting these differences in the UX to implement them properly.

**Comment by pastk** [comment]
- integrated latest main style changes (railway types, isolines)
- made tracks/paths/railways thinner
- increased visibility of cliffs, highway=service (not unusual for trailheads access)

Map not being updated after style switch on Android w/ Vulkan is fixed in #5775

**Comment by pastk** [comment]
- more visible national parks and viewpoints icons
- standardized font sizes for outdoors POIs
- reduced visibility of nameless peaks
- reduced visibilities of some POI labels

**Comment by pastk** [comment]
outdoors_2023-08_1

outdoors_2023-08_2

**Comment by pastk** [comment]
Higher dpi:

**Comment by pastk** [comment]
- increased rivers and streams visibility
- reduced natural=spring visibility to z12-
- isolines became too thick compared to paths/tracks, tuned down

**Comment by pastk** [comment]
Here is Mapy.cz vs AllTrails vs OM outdoors
(Komoot doesn't work for me, hangs at the registration attempt).

At a glance, we lack in overall map contrast and isolines detalisation (competitors have step 10, we have 50).

Mapy.cz is pretty similar otherwise, just different POIs have different visibilities from OM (need to zoom in more to see the water spring and the viewpoint in Mapy, need to zoom in to z17 in OM to see historic ruins).

Interestingly, AllTrails is pretty minimalistic: no water spring, cave, viewpoint, ruins - even when zoomed in all the way! Cliffs are present but displayed at higher zooms. And no differentiation for intermittent springs. So detail-wise we have an edge, but their map is more contrast and readable.

outdoorsOM_versus_others

**Comment by biodranik** [comment]
Komoot is using Mapbox styles. 
image

**Comment by pastk** [comment]
Rebased.

But to test it and make a new beta we need new maps @vng 

Alos increased alpine_hut visibility to z11-.

**Comment by RedAuburn** [comment]
https://www.thunderforest.com/maps/outdoors/preview/ in case you haven't seen it, thunderforest's style is really nice (similar to alltrails)

**Comment by pastk** [comment]
@vng @biodranik @rtsisyk 
Do you see any blockers to merge it into the master?

Note that its not about a full-blown "public release" (as we don't have an UI style switcher yet).

Its about giving an ability to use the style by enthusiast people (who don't mind typing `?olight` in the search to switch) already, making it easier to keep the style in sync with the main one, making current style contributors familiar with it and hopefully attracting more people to work on it.

**Comment by biodranik** [comment]
The main blocker for me is an ugly looking map.

**Comment by vng** [comment]
No problem on Core side, however, we can make (in future) a better geometry index, dividing not only by zooms, but also by style tags (Outdoor, Nav, ...).

**Comment by pastk** [comment]
> The main blocker for me is an ugly-looking map.

Could you please be more specific?
I'd be great to see some screenshots that looks ugly in your opinion and comparison with some other outdoors maps that look good in the same place/zoom.

**Comment by pastk** [comment]
Rebased.

Anyone who wants to build and test it - please use the [beta-outdoor](https://github.com/organicmaps/organicmaps/tree/beta-outdoor) branch (as it contains a commit with the latest outdoors-specific maps).

**Comment by pastk** [comment]
Some more comparisons with AllTrails.

Overview zooms (z12)
vs-AllTrails-z12

Note that we discard small lakes too early (our geometry optimization / simplification filters are very aggressive overall, so AllTrails has much smoother lines). They're also good in overall contrast, isolines & hillshading, highlighted roads. But lack in details like huts, small streams, water sources, vegetation cover (even an alpine tree line is not visible).

On z11 AllTrails stops displaying trails, etc. so not much useful.
We continue to display a lot of detail, which is useful in sparse areas but at the expense at more clutter in denser cases.
vs-AllTrails-z11_sm

In the city (an "uglinness" competition :))
vs-AllTrails-city

**Comment by patepelo** [comment]
Maybe for a next round, but have you considered representing these both features?

### Ridge
[Ridge](https://wiki.openstreetmap.org/wiki/Tag:natural%3Dridge )
In carto it looks like this:
image

### Arete
[Arete](https://wiki.openstreetmap.org/wiki/Tag:natural%3Darete)
In carto it looks like this:
image

**Comment by pastk** [comment]
Yeap for sure we'll add them! (e.g. we can use a cliff-like pattern, but double-sided).

ATM we lack many other `natural=` features like valleys, dunes, etc.
Many of ideas are mentioned in e.g.
- #2356
- #2804

**Comment by pastk** [comment]
> Let's start with fixing iOS first.

what's wrong with iOS here? did it stop working?

**Comment by biodranik** [comment]
Uups sorry, I forgot that it was already fixed. Neet to finish UI switches for both platform.

**Comment by pastk** [comment]
Let's merge it in a "for those who knows" mode first, nothing prevents it.
We'll add UI switching later - actually merging into master/release might attract devs who wants to work on it.

**Review by vng — APPROVED** [review]

**Pull request merged**